infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -175,231 +186,272 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
     }
 
     /**
-     * Perform a prefix trim.
+     * Perform a prefix trim (trim log).
      * Here the token is not used to perform the trim as the epoch at which the checkpoint was completed
      * might be old. Hence, we use the msg epoch to perform the trim. This should be safe provided that the
      * trim is performed only on the token provided by the CheckpointWriter which ensures that the checkpoint
      * was persisted. Using any other address to perform a trim can cause data loss.
      */
-    @ServerHandler(type = CorfuMsgType.PREFIX_TRIM)
-    private void prefixTrim(CorfuPayloadMsg<TrimRequest> msg, ChannelHandlerContext ctx,
-                            IServerRouter r) {
-        log.debug("prefixTrim: trimming prefix to {}", msg.getPayload().getAddress());
-        batchWriter
-                .addTask(PREFIX_TRIM, msg)
-                .thenRun(() -> r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg()))
+    @RequestHandler(type = PayloadCase.TRIM_LOG_REQUEST)
+    private void handleTrimLog(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        log.debug("handleTrimLog[{}]: trimming prefix to {}", req.getHeader().getRequestId(),
+                TextFormat.shortDebugString(req.getPayload().getTrimLogRequest().getAddress()));
+
+        batchWriter.addTask(BatchWriterOperation.Type.PREFIX_TRIM, req)
+                .thenRun(() -> {
+                    HeaderMsg header = getHeaderMsg(req.getHeader(), false, true);
+                    r.sendResponse(getResponseMsg(header, getTrimLogResponseMsg()), ctx);
+                })
                 .exceptionally(ex -> {
-                    handleException(ex, ctx, msg, r);
+                    handleException(ex, ctx, req, r);
                     return null;
                 });
     }
 
-  @ServerHandler(type = CorfuMsgType.READ_REQUEST)
-  public void read(CorfuPayloadMsg<ReadRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-    boolean cacheable = msg.getPayload().isCacheReadResult();
-    if (log.isTraceEnabled()) {
-      log.trace("read: {}, cacheable: {}", msg.getPayload().getAddresses(), cacheable);
-    }
+    @RequestHandler(type = PayloadCase.READ_LOG_REQUEST)
+    public void handleRead(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        final boolean cacheable = req.getPayload().getReadLogRequest().getCacheResults();
+        final List<Long> addressList = req.getPayload().getReadLogRequest().getAddressList();
 
-    ReadResponse rr = new ReadResponse();
+        if (log.isTraceEnabled()) {
+            log.trace("handleRead: {}, cacheable: {}", addressList, cacheable);
+        }
 
-    for (long address : msg.getPayload().getAddresses()) {
-      try {
-        ILogData logData = dataCache.get(address, cacheable);
-        if (logData == null) {
-          rr.put(address, LogData.getEmpty(address));
-        } else {
-          rr.put(address, (LogData) logData);
+        ReadResponse rr = new ReadResponse();
+
```

Comment:
This is not needed, you just unwrap it to `List<LogData> addresses`  and then construct a protobuf from that. Since a lot of quorum replication relies on that type, it would be more work to remove the `ReadResponse` class .